syntheseus/reaction_prediction/utils/downloading.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import os
+import urllib.request
+import zipfile
+from pathlib import Path
+from typing import Optional
+
+import yaml
+
+logger = logging.getLogger(__file__)
+
+
+def get_cache_dir(key: str) -> Path:
+    """Get the cache directory for a given key (e.g. model name)."""
+
+    # First, check if the cache directory has been manually overriden.
+    cache_dir_from_env = os.getenv("SYNTHESEUS_CACHE_DIR")
+    if cache_dir_from_env is not None:
+        # If yes, use the path provided.
+        cache_dir = Path(cache_dir_from_env)
+    else:
+        # If not, construct a reasonable default.
+        cache_dir = Path(os.getenv("HOME", ".")) / ".cache" / "torch" / "syntheseus"
+
+    cache_dir = cache_dir / key
+    cache_dir.mkdir(parents=True, exist_ok=True)
+
+    return cache_dir
+
+
+def get_cache_dir_download_if_missing(key: str, link: str) -> Path:
+    """Get the cache directory for a given key, but populate by downloading from link if empty."""
+
+    cache_dir = get_cache_dir(key)
+    if not any(cache_dir.iterdir()):
+        cache_zip_path = cache_dir / "model.zip"
+
+        logger.info(f"Downloading data from {link} to {cache_zip_path}")
+        urllib.request.urlretrieve(link, cache_zip_path)
+
+        with zipfile.ZipFile(cache_zip_path, "r") as f_zip:
+            f_zip.extractall(cache_dir)
+
+        cache_zip_path.unlink()
```

Comment:
Is there any reason why you don't use the `tempfile` package for this? Is it because you want to save the file between runs?

Reply:
Yes, exactly: downloading and unzipping can take at least a few seconds, and in some cases (e.g. RetroKNN with a 3GB checkpoint) several minutes. I wanted to make sure we only do this once. This also allows us to add more tests for the models into the integration CI pipeline without incurring this cost for every test.